tuner/tuner/candidate_gen.py

```diff
@@ -187,20 +187,232 @@ def __str__(self) -> str:
+                tile_size[idx] = k
+        return tile_size
+
+    @abstractmethod
+    def get_pipeline_config(self) -> str:
+        pass
+
+    @abstractmethod
+    def get_intrinsic_config(self) -> str:
+        pass
+
+    @abstractmethod
+    def apply_configuration(self, template: list[str], tile_sizes: list[int]) -> str:
+        pass
+
+
+@dataclass
+class LLVMGPUConfiguration(BaseConfiguration):
     gpu_pipeline_options: GpuPipelineOptions
     waves_per_eu: int
+    intrinsic: MfmaIntrinsic
+
+    def get_pipeline_config(self) -> str:
+        extra_config = ""
+        if not self.gpu_pipeline_options.all_default():
+            extra_config += f", gpu_pipeline_options = {self.gpu_pipeline_options}"
+        if self.waves_per_eu != 2:
+            extra_config += (
+                f', llvm_func_attrs = {{"amdgpu-waves-per-eu" = "{self.waves_per_eu}"}}'
+            )
+
+        return extra_config
+
+    def get_intrinsic_config(self) -> str:
+        return str(self.intrinsic)
+
+    def apply_configuration(self, template: list[str], tile_sizes: list[int]) -> str:
+        tune_logger.info(f"Applying: {self}")
+        expr0 = re.compile(
+            r"<intrinsic = #iree_gpu\.mma_layout<(.+)>, subgroup_m_count = ([0-9]+), subgroup_n_count = ([0-9]+)>"
+        )
+        expr1 = re.compile(
+            r"LLVMGPUVectorDistribute workgroup_size = \[.+\] subgroup_size = ([0-9]+),"
+        )
+        expr2 = re.compile(r"tile_sizes = \[\[([0-9]+)(, ([0-9]+))+\]\]")
+        expr3 = re.compile(
+            r"gpu_pipeline_options = #iree_gpu\.pipeline_options<([^>]*)>"
+        )
+        expr4 = re.compile(r"\"amdgpu-waves-per-eu\" = \"([0-9])\"")
+        repl0 = f"<intrinsic = #iree_gpu.mma_layout<{self.intrinsic}>, subgroup_m_count = {self.subgroup_m_count}, subgroup_n_count = {self.subgroup_n_count}>"
+        repl1 = f'LLVMGPUVectorDistribute workgroup_size = [{", ".join(map(str, self.workgroup_size))}] subgroup_size = {self.subgroup_size},'
+        repl2 = f'tile_sizes = [[{", ".join(map(str, tile_sizes))}]]'
+        repl3 = f"gpu_pipeline_options = {self.gpu_pipeline_options}"
+        repl4 = f'"amdgpu-waves-per-eu" = "{self.waves_per_eu}"'
+
+        new_mlir = ""
+        for line in template:
+            if "intrinsic =" in line:
+                line = re.sub(expr0, repl0, line)
+            if "LLVMGPUVectorDistribute " in line:
+                line = re.sub(expr1, repl1, line)
+            if "tile_sizes" in line:
+                line = re.sub(expr2, repl2, line)
+            if "gpu_pipeline_options =" in line:
+                line = re.sub(expr3, repl3, line)
+            if "amdgpu-waves-per-eu" in line:
+                line = re.sub(expr4, repl4, line)
+            new_mlir += line
+
+        return new_mlir
+
+
+@dataclass
+class LLVMCPUConfiguration(BaseConfiguration):
+    def get_pipeline_config(self) -> str:
+        return ""
+
+
+@dataclass
+class BaseConfigurationBuilder:
```

Comment:
Why do we need these builders?

Reply:
That was one of the only ways of creating the configurations that I could think of while trying to avoid hacks. I tried using a strategy pattern before, but it was slightly less clear than using builders + factory. It's still not a perfect approach though, because the `generate_solutions` method still loses some of the clarity. I'm considering moving that method inside of the configuration dataclasses and getting rid of the builder while keeping a factory for creating the configurations